### PR TITLE
Command line '-geometry WxH+X+Y' switch for X11/Unix

### PR DESCRIPTION
**Pull request by zelea2**
Width and Height are mandatory, X/Y position is optional
negative position numbers are distances between the right/bottom
edge of the window and right/bottom edge of the screen

### CODE REVIEW

**Comment by mattatobin** [comment]
Please change your commit message to follow the style in the contribution documentation.

File: toolkit/xre/nsAppRunner.cpp [hunk 1]
```diff
@@ -2815,11 +2816,22 @@ XREMain::XRE_mainInit(bool* aExitFlag)
   if (!home || !*home) {
     struct passwd *pw = getpwuid(geteuid());
     if (!pw || !pw->pw_dir) {
-      Output(true, "Could not determine HOME directory");
+      Output(true, "Could not determine HOME directory\n");
       return 1;
     }
     SaveWordToEnv("HOME", nsDependentCString(pw->pw_dir));
   }
+#  ifdef MOZ_X11
```
Comment (mattatobin) [inline toolkit/xre/nsAppRunner.cpp]: No, we don't do this indenting crap with preprocessor directives

File: toolkit/xre/nsAppRunner.cpp [hunk 2]
```diff
@@ -2815,11 +2816,22 @@ XREMain::XRE_mainInit(bool* aExitFlag)
...
+  const char *geometry;  
+  ar = CheckArg("geometry", false, &geometry );
+  if (ar == ARG_BAD) {
+    Output(true, "Error: geometry requieres an argument\n");
+    return 1;
+  }
+  nsCString window_geometry;
+  window_geometry.Assign(geometry);
+  SaveWordToEnv( "MOZILLA_WINDOW_GEOMETRY", window_geometry );
+#  endif
```
Comment (mattatobin) [inline toolkit/xre/nsAppRunner.cpp]: And no

File: xpfe/appshell/nsXULWindow.cpp [hunk 3]
```diff
@@ -1054,6 +1054,10 @@ void nsXULWindow::OnChromeLoaded()
     // trump that override.)
     if (!parentWindow)
       positionSet = false;
+    const char *val = PR_GetEnv("MOZILLA_WINDOW_GEOMETRY");
```
Comment (mattatobin) [inline xpfe/appshell/nsXULWindow.cpp]: This is not Linux specific it is unix but not macos

File: xpfe/appshell/nsXULWindow.cpp [hunk 4]
```diff
@@ -1129,8 +1133,11 @@ bool nsXULWindow::LoadPositionFromXUL(int32_t aSpecWidth, int32_t aSpecHeight)
   int32_t currY = 0;
   int32_t currWidth = 0;
   int32_t currHeight = 0;
+  int32_t screenWidth = 0;
```
Comment (mattatobin) [inline xpfe/appshell/nsXULWindow.cpp]: None of the rest of this is linux specific at all

File: toolkit/xre/nsAppRunner.cpp [hunk 5]
```diff
@@ -2815,11 +2816,22 @@ XREMain::XRE_mainInit(bool* aExitFlag)
...
+#  ifdef MOZ_X11    
+  const char *geometry;  
+  ar = CheckArg("geometry", false, &geometry );
+  if (ar == ARG_BAD) {
+    Output(true, "Error: geometry requieres an argument\n");
+    return 1;
+  }
+  nsCString window_geometry;
+  window_geometry.Assign(geometry);
+  SaveWordToEnv( "MOZILLA_WINDOW_GEOMETRY", window_geometry );
```
Comment (mattatobin) [inline toolkit/xre/nsAppRunner.cpp]: This is not the right naming scheme for env vars

File: xpfe/appshell/nsXULWindow.cpp [hunk 6]
```diff
@@ -1129,8 +1133,11 @@ bool nsXULWindow::LoadPositionFromXUL(int32_t aSpecWidth, int32_t aSpecHeight)
...
+  int32_t screenHeight = 0;
   nsresult errorCode;
-  int32_t temp;
+  int32_t temp, tp, tm, x, y;
+  bool    got_x, got_y, neg_x, neg_y;
```
Comment (mattatobin) [inline xpfe/appshell/nsXULWindow.cpp]: odd whitespace

File: xpfe/appshell/nsXULWindow.cpp [hunk 7]
```diff
@@ -1165,6 +1172,70 @@ bool nsXULWindow::LoadPositionFromXUL(int32_t aSpecWidth, int32_t aSpecHeight)
     gotPosition = true;
   }
 
+  got_x = got_y = neg_x = neg_y = false;
+  nsCString geometryString, part;
+  geometryString.Assign(PR_GetEnv("MOZILLA_WINDOW_GEOMETRY"));
```
Comment (mattatobin) [inline xpfe/appshell/nsXULWindow.cpp]: Improper env var name

File: xpfe/appshell/nsXULWindow.cpp [hunk 8]
```diff
@@ -1054,6 +1054,10 @@ void nsXULWindow::OnChromeLoaded()
     // trump that override.)
     if (!parentWindow)
       positionSet = false;
+    const char *val = PR_GetEnv("MOZILLA_WINDOW_GEOMETRY");
+    // allow window positioning if the geometry switch is set
+    if (val && *val)
```
Comment (mattatobin) [inline xpfe/appshell/nsXULWindow.cpp]: Improper code style

File: xpfe/appshell/nsXULWindow.cpp [hunk 9]
```diff
@@ -1165,6 +1172,70 @@ bool nsXULWindow::LoadPositionFromXUL(int32_t aSpecWidth, int32_t aSpecHeight)
     gotPosition = true;
   }
 
+  got_x = got_y = neg_x = neg_y = false;
+  nsCString geometryString, part;
+  geometryString.Assign(PR_GetEnv("MOZILLA_WINDOW_GEOMETRY")); 
+  tp = geometryString.FindChar( '+' );
+  tm = geometryString.FindChar( '-' );
+  temp = std::min(tp, tm);
+  if (temp < 0)
```
Comment (mattatobin) [inline xpfe/appshell/nsXULWindow.cpp]: Improper code style

File: xpfe/appshell/nsXULWindow.cpp [hunk 10]
```diff
@@ -1165,6 +1172,70 @@ bool nsXULWindow::LoadPositionFromXUL(int32_t aSpecWidth, int32_t aSpecHeight)
...
+    temp = std::max(tp, tm);
+  if (temp > 0) {
+    geometryString.Cut(0, temp + 1);
+    got_x = true;
+    if (temp == tm)
```
Comment (mattatobin) [inline xpfe/appshell/nsXULWindow.cpp]: Improper code style

File: xpfe/appshell/nsXULWindow.cpp [hunk 11]
```diff
@@ -1165,6 +1172,70 @@ bool nsXULWindow::LoadPositionFromXUL(int32_t aSpecWidth, int32_t aSpecHeight)
...
+      neg_x = true;
+  }
+  if( got_x )
```
Comment (mattatobin) [inline xpfe/appshell/nsXULWindow.cpp]: Screwy whitespace

File: xpfe/appshell/nsXULWindow.cpp [hunk 12]
```diff
@@ -1165,6 +1172,70 @@ bool nsXULWindow::LoadPositionFromXUL(int32_t aSpecWidth, int32_t aSpecHeight)
...
+      nsCOMPtr<nsIDOMScreen> screen = window->GetScreen();
+      if (screen) {
+        screen->GetAvailWidth(&screenWidth); 
+        screen->GetAvailHeight(&screenHeight);
+      }
+    }
+    tp = geometryString.FindChar('+');
+    tm = geometryString.FindChar('-');
+    temp = std::min(tp, tm);
+    if (temp < 0)
```
Comment (mattatobin) [inline xpfe/appshell/nsXULWindow.cpp]: Improper code style

File: xpfe/appshell/nsXULWindow.cpp [hunk 13]
```diff
@@ -1165,6 +1172,70 @@ bool nsXULWindow::LoadPositionFromXUL(int32_t aSpecWidth, int32_t aSpecHeight)
...
+      if (temp == tm) 
+        neg_y = true;
+    }
+    if (got_y) 
+      geometryString.Left(part,temp);
+    else
+      part = geometryString;
+    x = part.ToInteger(&errorCode);
+    if (NS_SUCCEEDED(errorCode)) {
+      if (neg_x)
```
Comment (mattatobin) [inline xpfe/appshell/nsXULWindow.cpp]: Improper code style

File: xpfe/appshell/nsXULWindow.cpp [hunk 14]
```diff
@@ -1165,6 +1172,70 @@ bool nsXULWindow::LoadPositionFromXUL(int32_t aSpecWidth, int32_t aSpecHeight)
...
+      geometryString.Left(part,temp);
+    else
+      part = geometryString;
+    x = part.ToInteger(&errorCode);
+    if (NS_SUCCEEDED(errorCode)) {
+      if (neg_x) 
+        specX = screenWidth - aSpecWidth - x;
+      else
+        specX = x;
+      if (specX >= 0)
```
Comment (mattatobin) [inline xpfe/appshell/nsXULWindow.cpp]: Improper code style

File: xpfe/appshell/nsXULWindow.cpp [hunk 15]
```diff
@@ -1165,6 +1172,70 @@ bool nsXULWindow::LoadPositionFromXUL(int32_t aSpecWidth, int32_t aSpecHeight)
...
+        gotPosition = true;
+    }
+    if (got_y)
```
Comment (mattatobin) [inline xpfe/appshell/nsXULWindow.cpp]: Improper code style

File: xpfe/appshell/nsXULWindow.cpp [hunk 16]
```diff
@@ -1165,6 +1172,70 @@ bool nsXULWindow::LoadPositionFromXUL(int32_t aSpecWidth, int32_t aSpecHeight)
...
+        specX = x;
+      if (specX >= 0)
+        gotPosition = true;
+    }
+    if (got_y) 
+    {
+      geometryString.Cut(0, temp + 1);
+      y = geometryString.ToInteger(&errorCode);
+      if (NS_SUCCEEDED(errorCode)) {
+        if (neg_y)
```
Comment (mattatobin) [inline xpfe/appshell/nsXULWindow.cpp]: I'm done.

**Comment by mattatobin** [comment]
Tag Issue #1466

File: toolkit/xre/nsAppRunner.cpp [hunk 17]
```diff
@@ -2815,11 +2816,22 @@ XREMain::XRE_mainInit(bool* aExitFlag)
   if (!home || !*home) {
     struct passwd *pw = getpwuid(geteuid());
     if (!pw || !pw->pw_dir) {
-      Output(true, "Could not determine HOME directory");
+      Output(true, "Could not determine HOME directory\n");
```
Comment (mattatobin) [inline toolkit/xre/nsAppRunner.cpp]: Unrelated change.

**Comment by wolfbeast** [comment]
I'm closing this PR since it obviously needs redone. Considering the complexity involved in parsing this parameter, I also request that in a retry, you isolate all this in its own parsing function with proper error checking and success/failure return with the callsite only accepting the parameter if it passes all checks and ignoring the parameter otherwise.

File: xpfe/appshell/nsXULWindow.cpp (hunk 15 above)
Comment (zelea2) [reply xpfe/appshell/nsXULWindow.cpp]: A lot of your issues that you list are related to missing curly brackets around single line if/else.
In addition to these changes can you provide me an ".indent.pro" so I can run my code through indent and get a consistent code style for this project?

File: toolkit/xre/nsAppRunner.cpp (hunk 17 above)
Comment (zelea2) [reply toolkit/xre/nsAppRunner.cpp]: It is unrelated but that LF is needed.

File: xpfe/appshell/nsXULWindow.cpp (hunk 7 above)
Comment (zelea2) [reply xpfe/appshell/nsXULWindow.cpp]: What is a proper environment name?

File: xpfe/appshell/nsXULWindow.cpp (hunk 15 above)
Comment (mattatobin) [reply xpfe/appshell/nsXULWindow.cpp]: I don't know what that is and I would suggest staying away from any automatic formatters.

The entire patch needs redone.

Comment (zelea2) [reply xpfe/appshell/nsXULWindow.cpp]: Why should I stay away from formatters? I am not gone change my programming style every time I contribute to a different project. Anyway I'll figure that myself.

You are not very helpful calling my patch crap without suggesting improvements.
wolfbeast at least said what needs to be changed  and I'll do the sanity checking.

I have also asked you how to chose a "proper" environment variable name.

Comment (wolfbeast) [reply xpfe/appshell/nsXULWindow.cpp]: There is a very god reason I've been spending considerable time creating a coding style document. USE IT.
See `/docs` in the source tree

**Comment by mattatobin** [comment]
@zelea2 We aren't going to argue with you on this. If you have additional questions you can direct them to the Issue but this PR has been rejected and is over. You were given instructions if you want to try again.
